sympy/solvers/solveset.py

```diff
@@ -353,39 +354,113 @@ def _is_function_class_equation(func_class, f, symbol):
+    g, h = g.expand(), h.expand()
     if not h.has(symbol):
         return _solve_as_poly(g, symbol, domain)
     else:
-        valid_solns = _solveset(g, symbol, domain)
-        invalid_solns = _solveset(h, symbol, domain)
-        return valid_solns - invalid_solns
+        # search for exp(I*x) terms
+        flag = False
+        atoms = list(g.atoms(exp))
+        atoms += list(h.atoms(exp))
+        for atom in atoms:
+            for img in expand_log(atom, force=True).atoms(I):
+                if img:
+                    flag = True
+        if flag:
+            # to get soln in desired form  for
+            # solveset((tan(x)-1).rewrite(exp), x, S.Reals)
+            # and these type of expressions, we need to
+            # make polynomial equation by replace I*symbol with dummy.
+            y = Dummy('y')
+            # make it independent from exp(I*symbol)
+            g, h = g.subs(exp(I*symbol), y), h.subs(exp(I*symbol), y)
+
+            solns = _solveset(g, y, domain) - _solveset(h, y, domain)
+            if g.has(symbol) or h.has(symbol):
+                return ConditionSet(symbol, Eq(f, 0), S.Reals)
+            if isinstance(solns, FiniteSet):
+                result = Union(*[invert_complex(exp(I*symbol), s, symbol)[1]
+                       for s in solns])
+                return result.intersection(domain) \
+                if domain.is_subset(S.Reals) else result
+            elif solns is S.EmptySet:
+                return S.EmptySet
+            else:
+                return ConditionSet(symbol, Eq(f_original, 0), S.Reals)
+        else:
+            valid_solns = _solveset(g, symbol, domain)
+            invalid_solns = _solveset(h, symbol, domain)
+            return valid_solns - invalid_solns
 
 
 def _solve_trig(f, symbol, domain):
     """ Helper to solve trigonometric equations """
-    f = trigsimp(f)
-    f_original = f
-    f = f.rewrite(exp)
-    f = together(f)
-    g, h = fraction(f)
-    y = Dummy('y')
-    g, h = g.expand(), h.expand()
-    g, h = g.subs(exp(I*symbol), y), h.subs(exp(I*symbol), y)
-    if g.has(symbol) or h.has(symbol):
-        return ConditionSet(symbol, Eq(f, 0), S.Reals)
+    from sympy import factor, factor_list
+
+    def good_for_factor(trig_eq):
+        i = 0
+        for a in trig_eq.args:
+            if isinstance(a, TrigonometricFunction) or \
+                                isinstance(a, HyperbolicFunction):
+                i += 1
+                if i >= 2:
+                    return True
+
+    if _is_function_class_equation(TrigonometricFunction, f, symbol):
+        f = trigsimp(f)
+    # trigsimp is not defined for hyperbolic
+    # TODO trigh function for Hyperbolic Functions if required
+    f_orig = f
+    # factor_list isn't working for factor_list(sqrt(2)*sin(x))
+    # so `.as_independent` is used. Fix issue => remove below line.
+    # issue #11198
+    if f.is_Mul:
+        ind, f = f.as_independent(symbol)
+    fact_list = factor_list(f)[1]
```

Comment:
` factor` and `factor_list` is used in correct order to get simple solution: [gist](https://gist.github.com/Shekharrajak/17fdcd2320f572fc9fc8674823137e20)